gc/base/standard/Scavenger.cpp

```diff
@@ -1723,30 +1723,28 @@ MM_Scavenger::deepScanOutline(MM_EnvironmentStandard *env, omrobjectptr_t object
 	env->_scavengerStats._totalDeepStructures += 1;
 #endif /* J9MODRON_TGC_PARALLEL_STATISTICS */
 
-	while (true) {
-		GC_SlotObject tempSlot(env->getOmrVM(), (fomrobject_t*)(((uintptr_t) tempObj) + priorityField));
-		if (NULL == tempSlot.readReferenceFromSlot()) {
-			if ((priorityField == selfReferencingField2) || (selfReferencingField2 == 0)) {
-				break;
-			}
-			priorityField = selfReferencingField2;
-		} else {
-			copyAndForward(env, &tempSlot);
-			/* Did we encounter an already visited object or hit throttling threshold? */
-			if (NULL == env->_effectiveCopyScanCache) {
+	do {
+		GC_SlotObject prioritySlot(env->getOmrVM(), (fomrobject_t*)(((uintptr_t) currentDeepObj) + priorityField));
+		copyAndForward(env, &prioritySlot);
+		/* Did we encounter an already visited/NULL object? */
+		if (NULL == env->_effectiveCopyScanCache) {
+			/* Can't continue any further - attempt to deep scan with other self referencing field (e.g, prev field) */
+			if ((priorityField == priorityFieldOffset2) || (priorityFieldOffset2 == 0)) {
 				break;
 			}
+			priorityField = priorityFieldOffset2;
+			continue;
+		}
 
 #if defined(J9MODRON_TGC_PARALLEL_STATISTICS)
-			objDeepScanned += 1;
+		objDeepScanned += 1;
 #endif /* J9MODRON_TGC_PARALLEL_STATISTICS */
 
-			if(env->approxScanCacheCount > freeListUtilizationLimit){
-				break;
-			}
-			tempObj = tempSlot.readReferenceFromSlot();
+		if(env->approxScanCacheCount > freeListUtilizationLimit) {
+			break;
 		}
-	}
+		currentDeepObj = prioritySlot.readReferenceFromSlot();
```

Comment:
Put a comment just before this along these lines:  even though we just successfully copied an object and its new address is normally could be obtained from reading prioritySlot, in CS world a mutator thread could have raced and overwritten the slot with NULL (or even some other non-NULL value), hence we need a NULL point check before we proceed with another iterator.